pkg/upgrade/config.go

```diff
@@ -44,8 +44,36 @@ func (t *TargetClusterConfig) UpgradeScopes() []string {
 
 // KeyPairConfig is something
 type KeyPairConfig struct {
-	SecretRef    string `json:"secretRef,omitempty"`
-	ClusterField string `json:"clusterField,omitempty"`
+	SecretRef           string `json:"secretRef,omitempty"`
+	ClusterField        string `json:"clusterField,omitempty"`
+	KubeconfigSecretRef string `json:"kubeconfigSecretRef,omitempty"`
+
+	// APIEndpoint is a URL to a kube-apiserver.
+	// This entry is used only if SecretRef or ClusterField is set.
+	// APIEndpoint is ignored if KubeconfigSecretRef is set.
+	APIEndpoint string `json:"apiEndpoint"`
```

Comment:
I moved this into the KeyPairConfig object because this is the object it is used with/required for.

There is also an argument for keeping it in TargetClusterConfig and moving the following validation function onto the TargetClusterClient. I don't really care either way, this just made the most sense to me at the time of writing.